### PR TITLE
Reimplement workqueue.NewItemBucketRateLimiter to batch requests.

### PR DESCRIPTION
**Pull request by mborsz**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

This reimplements a NewItemBucketRateLimiter to implement batching inside of rate limiter instead of checking if given element is already in queue (as proposed in https://github.com/kubernetes/kubernetes/pull/88579). This way we can get rid of `DeduplicatingRateLimitingQueue`, `notStarted` map and simplify implementation significantly.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

This is intended to be used in https://github.com/kubernetes/kubernetes/pull/88161 instead of data structure created in https://github.com/kubernetes/kubernetes/pull/88579

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

/assign @lavalamp 
/assign @freehan 
/cc @mm4tt

### CODE REVIEW

File: staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go [hunk 1]
```diff
@@ -62,24 +62,57 @@ func (r *BucketRateLimiter) NumRequeues(item interface{}) int {
...
 
 // NewItemBucketRateLimiter creates new ItemBucketRateLimiter instance.
-func NewItemBucketRateLimiter(r rate.Limit, burst int) *ItemBucketRateLimiter {
+func NewItemBucketRateLimiter(qps float64, burst int) *ItemBucketRateLimiter {
 	return &ItemBucketRateLimiter{
-		r:        r,
+		qps:      qps,
 		burst:    burst,
-		limiters: make(map[interface{}]*rate.Limiter),
+		limiters: make(map[interface{}]*itemState),
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go]: What keeps this from growing without bound?

File: staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters_test.go [hunk 2]
```diff
@@ -99,18 +97,24 @@ func TestItemExponentialFailureRateLimiterOverFlow(t *testing.T) {
...
+		if got := limiter.When("one"); got == 0 {
+			t.Errorf("limiter.When(one) = %v; want > 0", got)
+		}
 	}
-	// limiter.When should be at this point = 1000 * rate.Limit.
-	// We set the threshold 1s below this value to avoid race conditions.
-	if got, want := limiter.When("one"), 990*100*time.Millisecond; got < want {
+	// limiter.When should be at this point = 1s / 10.0 = 100 ms.
+	// We check of 90ms to avoid races.
+	if got, want := limiter.When("one"), 90*time.Millisecond; got < want {
```
Comment (lavalamp) [inline staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters_test.go]: Use the clock package to not depend on wall time and remove that source of raciness.

I am surprised this passes, I would have expected `got` to be very close to 100ms, certainly not less than 90ms. Please bound it on both sides to prove that exactly one additional batch is produced by this test.

File: staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go (hunk 1 above)
Comment (mm4tt) [reply staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go]: That's a really good point. It definitely needs to be addressed

/lgtm cancel

Comment (mborsz) [reply staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go]: The user of this class is supposed to call Forget() when given key no longer exists in the system.

Forget() removes objects from this map.

In this case, we should keep only a single itemState object for only for all keys (services) that still exists in the system.

**Comment by mborsz** [comment]
I'm closing this PR as we decided to move forward with a simpler approach that doesn't need this: https://github.com/kubernetes/kubernetes/pull/88745

**Pull request closed without merging**